pkg/config/dynamicloader.go

```diff
@@ -0,0 +1,495 @@
+			if singletonCheck[cfg.Name()] > 1 {
+				// TODO (mattdurham) would be nice if it said where they were found
+				returnError = multierror.Append(returnError, fmt.Errorf("found multiple instances of singleton integration %s", cfg.Name()))
+			}
+		}
+	}
+	return configs, returnError
+}
+
+func (c *DynamicLoader) processLogs() (*logs.Config, error) {
+	filter := "logs-*.yml"
+	if c.cfg.LogsFilter != "" {
+		filter = c.cfg.LogsFilter
+	}
+
+	for _, path := range c.cfg.TemplatePaths {
+		result, err := c.generateConfigsFromPath(path, filter, func() interface{} {
+			return &logs.Config{}
+		}, c.handleMatch)
+		if err != nil {
+			return nil, err
+		}
+		if len(result) > 1 {
+			return nil, errors.New("multiple logs templates found")
+		}
+		if len(result) == 1 {
+			return result[0].(*logs.Config), nil
+		}
+
+	}
+	return nil, nil
+}
+
+// processTraces will return the traces configuration following the pattern `traces-*.yml`
+func (c *DynamicLoader) processTraces() (*traces.Config, error) {
+	filter := "traces-*.yml"
+	if c.cfg.TracesFilter != "" {
+		filter = c.cfg.TracesFilter
+	}
+
+	for _, path := range c.cfg.TemplatePaths {
+		result, err := c.generateConfigsFromPath(path, filter, func() interface{} {
+			return &traces.Config{}
+		}, c.handleMatch)
+		if err != nil {
+			return nil, err
+		}
+		if len(result) > 1 {
+			return nil, errors.New("multiple traces templates found")
+		}
+		if len(result) == 1 {
+			return result[0].(*traces.Config), nil
+		}
+
+	}
+	return nil, nil
+}
+
+// generateConfigsFromPath creates a series of yaml configs based on the path and a given string pattern.
+// the pattern is the same as used by filepath.Match.
+func (c *DynamicLoader) generateConfigsFromPath(path string, pattern string, configMake func() interface{}, matchHandler func(fs.FS, fs.DirEntry, func() interface{}) ([]interface{}, error)) ([]interface{}, error) {
+	handler, err := c.mux.Lookup(path)
+	if err != nil {
+		return nil, err
+	}
+	files, err := fs.ReadDir(handler, ".")
+	if err != nil {
+		return nil, err
+	}
+	var configs []interface{}
+	for _, f := range files {
+		// We don't recurse into directories
+		if f.IsDir() {
+			continue
+		}
+		matched, err := filepath.Match(pattern, f.Name())
+		if err != nil {
+			return nil, err
+		}
+		if matched {
+			matchedConfigs, err := matchHandler(handler, f, configMake)
+			if err != nil {
+				return nil, err
+			}
+			configs = append(configs, matchedConfigs...)
+		}
+	}
+	return configs, nil
+}
+
+func (c *DynamicLoader) handleAgentMatch(handler fs.FS, f fs.DirEntry, configMake func() interface{}) ([]interface{}, error) {
+	file, err := handler.Open(f.Name())
+	if err != nil {
+		return nil, err
+	}
+	stats, err := f.Info()
+	if err != nil {
+		return nil, err
+	}
+	fBytes := make([]byte, stats.Size())
+	bytesRead, err := file.Read(fBytes)
+	if bytesRead == 0 {
+		return nil, errors.New("no bytes read")
+	}
+
+	if err != nil && err != io.EOF {
+		return nil, err
+	}
```

Comment:
All of this can be replaced with a call to fs.ReadFile :) 

```suggestion
	fBytes, err := fs.ReadFile(handler, f.Name())
	if err != nil {
		return nil, err	
	}  
```

(ditto for the other handlers)

Reply:
Jeez, yeah thats a ton cleaner.